util/dsmetool.c

```diff
@@ -1,413 +1,532 @@
-"                                   SHUTDOWN USER ACTDEAD REBOOT BOOT NOT_SET\n"
-"                                   TEST MALF LOCAL or UNKNOWN\n"
-"  -h --help                       Print usage\n");
+        int64_t res = 0;
+
+        struct timespec ts;
+
+        if( clock_gettime(CLOCK_BOOTTIME, &ts) == 0 ) {
+                res = ts.tv_sec;
+                res *= 1000;
+                res += ts.tv_nsec / 1000000;
+        }
+
+        return res;
 }
 
-static void connect_to_dsme(void)
+static const char *dsme_msg_type_repr(int type)
 {
-    if( conn != 0 ) {
-        /* If we ever get here, there is something
-         * wrong with logic somewhere ... */
-        fprintf(stderr, "Double connect detected\n");
-        exit(EXIT_FAILURE);
-    }
-    conn = dsmesock_connect();
-    if (conn == 0) {
-        perror("dsmesock_connect");
-        exit(EXIT_FAILURE);
+#define X(name,value) if( type == value ) return #name
+
+    // public
+    X(CLOSE,                        0x00000001);
+    X(STATE_CHANGE_IND,             0x00000301);
+    X(STATE_QUERY,                  0x00000302);
+    X(SAVE_DATA_IND,                0x00000304);
+    X(POWERUP_REQ,                  0x00000305);
+    X(SHUTDOWN_REQ,                 0x00000306);
+    X(SET_ALARM_STATE,              0x00000307);
+    X(REBOOT_REQ,                   0x00000308);
+    X(STATE_REQ_DENIED_IND,         0x00000309);
+    X(THERMAL_SHUTDOWN_IND,         0x00000310);
+    X(SET_CHARGER_STATE,            0x00000311);
+    X(SET_THERMAL_STATE,            0x00000312);
+    X(SET_EMERGENCY_CALL_STATE,     0x00000313);
+    X(SET_BATTERY_STATE,            0x00000314);
+    X(BATTERY_EMPTY_IND,            0x00000315);
+    X(PROCESSWD_CREATE,             0x00000500);
+    X(PROCESSWD_DELETE,             0x00000501);
+    X(PROCESSWD_CLEAR,              0x00000502);
+    X(PROCESSWD_SET_INTERVAL,       0x00000503);
+    X(PROCESSWD_PING,               0x00000504);
+    X(PROCESSWD_PONG,               0x00000504);
+    X(PROCESSWD_MANUAL_PING,        0x00000505);
+    X(WAIT,                         0x00000600);
+    X(WAKEUP,                       0x00000601);
+    X(GET_VERSION,                  0x00001100);
+    X(DSME_VERSION,                 0x00001101);
+    X(SET_TA_TEST_MODE,             0x00001102);
+
+    //internal
+    X(DBUS_CONNECT,                 0x00000100);
+    X(DBUS_DISCONNECT,              0x00000101);
+    X(SHUTDOWN,                     0x00000316);
+    X(SET_USB_STATE,                0x00000317);
+    X(TELINIT,                      0x00000318);
+    X(CHANGE_RUNLEVEL,              0x00000319);
+    X(HEARTBEAT,                    0x00000702);
+    X(ENTER_MALF,                   0x00000900);
+    X(SET_LOGGING_VERBOSITY,        0x00001103);
+    X(IDLE,                         0x00001337);
+    X(DISK_SPACE,                   0x00002000);
+
+#undef X
+
+    return "UNKNOWN";
+}
+
+static const char *dsme_state_repr(dsme_state_t state)
+{
```

Comment:
dsme_state_repr and dsme_msg_type_repr are going very similar things but in very different ways. Looks pretty strange.